patches/chromium/ozone_wayland_add_support_for_content_insets.patch

```diff
@@ -0,0 +1,253 @@
+   auto old_scale = window_scale();
+   WaylandWindow::UpdateWindowScale(update_bounds);
+@@ -305,7 +333,8 @@ void WaylandToplevelWindow::HandleToplevelConfigure(int32_t width,
+   // explicitly set the bounds to the current desired ones or the previous
+   // bounds.
+   if (width > 1 && height > 1) {
+-    pending_bounds_dip_ = gfx::Rect(0, 0, width, height);
++    pending_bounds_dip_ = gfx::Rect(0, 0, width + content_insets_.width(),
++                                    height + content_insets_.height());
+   } else if (is_normal) {
+     pending_bounds_dip_.set_size(
+         gfx::ScaleToRoundedSize(GetRestoredBoundsInPixels().IsEmpty()
+@@ -342,7 +371,7 @@ void WaylandToplevelWindow::HandleSurfaceConfigure(uint32_t serial) {
+     //     change window size will not cause the window to redraw.
+     // Hence, acknowledge this |configure| now to tell the Wayland compositor
+     // that this window has been configured.
+-    shell_toplevel()->SetWindowGeometry(pending_bounds_dip_);
++    UpdateWindowGeometry(pending_bounds_dip_);
+     shell_toplevel()->AckConfigure(serial);
+     connection()->ScheduleFlush();
+   } else if (!pending_configures_.empty() &&
+@@ -379,7 +408,7 @@ void WaylandToplevelWindow::UpdateVisualSize(const gfx::Size& size_px) {
+                    });
+ 
+   if (result != pending_configures_.end()) {
+-    shell_toplevel()->SetWindowGeometry(gfx::Rect(size_dip));
++    UpdateWindowGeometry(gfx::Rect(size_dip));
+     shell_toplevel()->AckConfigure(result->serial);
+     connection()->ScheduleFlush();
+     pending_configures_.erase(pending_configures_.begin(), ++result);
+@@ -579,6 +608,11 @@ void WaylandToplevelWindow::SetOrResetRestoredBounds() {
+   }
+ }
+ 
++void WaylandToplevelWindow::UpdateWindowGeometry(gfx::Rect bounds) {
++  bounds.Inset(content_insets_);
++  shell_toplevel()->SetWindowGeometry(bounds);
++}
++
+ void WaylandToplevelWindow::SetUpShellIntegration() {
+   // This method should be called after the XDG surface is initialized.
+   DCHECK(shell_toplevel_);
+diff --git a/ui/ozone/platform/wayland/host/wayland_toplevel_window.h b/ui/ozone/platform/wayland/host/wayland_toplevel_window.h
+index 1282e995a36118a3491a42c453c53ee73cb1ff21..94ea15254e1263253546d60b8c5614835da1980c 100644
+--- a/ui/ozone/platform/wayland/host/wayland_toplevel_window.h
++++ b/ui/ozone/platform/wayland/host/wayland_toplevel_window.h
+@@ -7,6 +7,7 @@
+ 
+ #include "base/containers/circular_deque.h"
+ #include "build/chromeos_buildflags.h"
++#include "ui/gfx/geometry/insets.h"
+ #include "ui/gfx/geometry/vector2d.h"
+ #include "ui/ozone/platform/wayland/host/wayland_window.h"
+ #include "ui/platform_window/extensions/wayland_extension.h"
+@@ -35,6 +36,8 @@ class WaylandToplevelWindow : public WaylandWindow,
+   // be called after processing all pending events in the wayland connection.
+   void ApplyPendingBounds();
+ 
++  void UpdateContentInsets(const gfx::Insets& insets);
```

Comment:
Why is Chromium not setting content insets itself? Do PWA apps of Chrome browser work correctly under Wayland CSD?

Reply:
@zcbenz Yes but actually no...it depends on your definition of "work correctly".

If you're on Xorg and using CSD, Chromium does not use any shadows at all, instead opting for a rather large and [incredibly unpopular 3D border effect](https://bugs.chromium.org/p/chromium/issues/detail?id=920768) to "mimic" the look. If you're on Wayland, and you switch the window decorations to look native, the title bar has no borders or shadows at all and also looks rather terrible.

As of right now, I am not sure if this will ever be fixed upstream. I had hopes that LaCrOS would result in this being fixed...but it appears that they've just opted for the shadows to be provided by the Chrome OS compositor rather than done by the client, unfortunately, meaning there's no currently available workaround for native Linux.